File: examples/lua/src/client/todos.client.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local remotes = require(ReplicatedStorage.Shared.remotes)
local names = require(ReplicatedStorage.Shared.names)

local todos: { string } = {}

local function addTodo()
	remotes.addTodo:fire(names.getRandomName())
end

local function removeTodo()
	if #todos > 0 then
		remotes.removeTodo:fire(todos[math.random(#todos)])
	end
end

remotes.getTodos:request():andThen(function(serverTodos)
	todos = serverTodos
	print("New todos:", table.concat(todos, ", "))

	for _ = 1, 3 do
		addTodo()
	end
end)

remotes.todosChanged:connect(function(serverTodos)
	todos = serverTodos
	print("New todos:", table.concat(todos, ", "))
end)

while true do
	task.wait(math.random(1, 3))

	if math.random() > 0.5 then
		addTodo()
	else
		removeTodo()
	end
end

File: examples/ts/src/client/todos.client.ts
import { getRandomName } from "shared/names";
import { remotes } from "shared/remotes";

let todos: readonly string[] = [];

function addTodo() {
	remotes.addTodo.fire(getRandomName());
}

function removeTodo() {
	if (!todos.isEmpty()) {
		const index = math.random(0, todos.size() - 1);
		remotes.removeTodo.fire(todos[index]);
	}
}

remotes.getTodos.request().then((newTodos) => {
	todos = newTodos;
	print(`New todos: ${newTodos.join(", ")}`);

	for (const _ of $range(1, 3)) {
		addTodo();
	}
});

remotes.todosChanged.connect((newTodos) => {
	todos = newTodos;
	print(`New todos: ${newTodos.join(", ")}`);
});

// eslint-disable-next-line no-constant-condition
while (true) {
	task.wait(math.random(1, 3));

	if (math.random() > 0.5) {
		addTodo();
	} else {
		removeTodo();
	}
}
